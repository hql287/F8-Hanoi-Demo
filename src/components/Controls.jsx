// React Libraries
import path from 'path';
import React, {Component} from 'react';
import styled from 'styled-components';

export const Dashboard = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  border-radius: 10px;
  padding: 20px;
  button {
    margin: 0 5px;
  }
`;

export const Button = styled.button`

`;

class Controls extends Component {
  render() {
    const { jump, moveLeft, moveRight, toggleLabels } = this.props;
    return (
      <Dashboard>
        <Button onClick={moveLeft}>← Move Left</Button>
        <Button onClick={jump}>↑ Jump</Button>
        <Button onClick={moveRight}>Move Right →</Button>
        <Button onClick={toggleLabels}>Toogle Labels</Button>
      </Dashboard>
    );
  }
}

export default Controls;

