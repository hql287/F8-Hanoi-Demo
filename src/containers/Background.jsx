// React Libraries
import React from 'react';
import Label from '../components/Label';
import path from 'path';

const tilePath = path.resolve(__dirname, './static/imgs/tile_small.png');
const pipePath = path.resolve(__dirname, './static/imgs/pipe.png');

import styled from 'styled-components';

export const Layout = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
`;

export const Sky = styled.div`
  flex: 1;
  background: lightblue;
`;

export const Ground = styled.div`
  flex: 1;
  background: url(${tilePath});
  ${props =>
    props.groundHeight &&
    `
    max-height: ${props.groundHeight}px;
  `};
  ${props =>
    props.borderHeight &&
    `
    border-top: ${props.borderHeight}px solid darkgreen;
  `};
`;

function Background(props) {
  return (
    <Layout>
      <Sky>
        { props.showLabels &&
          <Label text="Sky" />
        }
      </Sky>
      <Ground {...props}>
        { props.showLabels &&
          <Label text="Ground" />
        }
      </Ground>
    </Layout>
  );
}

Background.defaultProps = {
  showLabels: false,
  groundHeight: 240,
  borderHeight: 10,
};

export default Background;
