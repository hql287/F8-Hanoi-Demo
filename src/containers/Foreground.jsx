import React from 'react';
import Character from '../components/Character';

import styled from 'styled-components';
export const Layout = styled.div`
  position: relative;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  height: 100%;
  width: 100%;
  flex-direction: column;
`;

function Foreground(props) {
  return (
    <Layout>
      <Character {...props} />
    </Layout>
  );
}

export default Foreground;
