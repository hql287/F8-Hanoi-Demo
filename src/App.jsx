// React Libraries
import React, {Fragment, Component} from 'react';
import styled, {injectGlobal} from 'styled-components';

import Controls from './components/Controls';
import Foreground from './containers/Foreground';
import Background from './containers/Background';

injectGlobal`
  body {
    margin: 0;
    background: #fbfbfb;
    height: 100vh;
    width: 100vw;
  }
  #app {
    height: 100vh;
    width: 100vw;
  }
  * {
    box-sizing: border-box;
  }
`;

const stepDistance = 10;
const borderHeight = 10;
const groundHeight = 130;

export default class App extends Component {
  constructor(props) {
    super(props);
    this.jump = this.jump.bind(this);
    this.moveLeft = this.moveLeft.bind(this);
    this.moveRight = this.moveRight.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.toggleLabels = this.toggleLabels.bind(this);
    this.state = {
      pos: {
        x: -300,
        y: 0,
      },
      showLabels: true,
    }
  }

  handleKeyDown(event) {
    const key = event.keyCode;
    // console.log('key: ', key);
    if (key === 37) this.moveLeft();
    if (key === 39) this.moveRight();
    return;
  }

  jump() {
    console.log('Jump');
  }

  toggleLabels() {
    this.setState({
      showLabels: !this.state.showLabels
    });
  }

  moveLeft() {
    this.setState({
      pos: Object.assign({}, this.state.pos, {
        x: this.state.pos.x - stepDistance,
      })
    })
  }

  moveRight() {
    this.setState({
      pos: Object.assign({}, this.state.pos, {
        x: this.state.pos.x + stepDistance,
      })
    })
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleKeyDown, false);
  }

  componentWillUnMount() {
    document.removeEventListener('keydown', this.handleKeyDown, false);
  }

  render() {
    return (
      <Fragment>
        <Background
          showLabels={this.state.showLabels}
          groundHeight={groundHeight}
          borderheight={borderHeight} />
        <Foreground
          showLabels={this.state.showLabels}
          groundHeight={groundHeight}
          borderHeight={borderHeight}
          pos={this.state.pos}
        />
        <Controls
          jump={this.jump}
          moveLeft={this.moveLeft}
          moveRight={this.moveRight}
          toggleLabels={this.toggleLabels}
        />
      </Fragment>
    );
  }
}
