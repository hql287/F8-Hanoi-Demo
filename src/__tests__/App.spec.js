// Libs
import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

// Component to test
import App from '../App';
import Background from '../containers/Background';
import Foreground from '../containers/Foreground';
import Controls from '../components/Controls';

describe('Renders the Background correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should render a Background component', () => {
    expect(wrapper.find(Background)).toHaveLength(1);
    expect(wrapper.find(Background)).not.toHaveLength(2);
  });

  it('should render a Foreground component', () => {
    expect(wrapper.find(Foreground)).toHaveLength(1);
    expect(wrapper.find(Foreground)).not.toHaveLength(2);
  });

  it('should render a Controls component', () => {
    expect(wrapper.find(Controls)).toHaveLength(1);
    expect(wrapper.find(Controls)).not.toHaveLength(2);
  });

  // Snapshot Testing
  it('should match snapShot', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
