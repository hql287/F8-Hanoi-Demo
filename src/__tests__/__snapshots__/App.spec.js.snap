// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the Background correctly to the DOM should match snapShot 1`] = `
Array [
  <div
    className="Background__Layout-iLolKP cVmhBA"
  >
    <div
      className="Background__Sky-jUevNA ljmzoM"
    >
      <p
        className="Label__LabelStyle-dONxjx iUOqgJ"
      >
        <p>
          Sky
        </p>
      </p>
    </div>
    <div
      className="Background__Ground-kryFUJ AewKv"
    >
      <p
        className="Label__LabelStyle-dONxjx iUOqgJ"
      >
        <p>
          Ground
        </p>
      </p>
    </div>
  </div>,
  <div
    className="Foreground__Layout-ieRzSl bKizyj"
  >
    <div
      className="Character__Layout-bdFJdW kOzMK"
    >
      <img
        alt=""
        className="Character__Speech-eSrpZz eBgjjc"
        src="/Users/quochungle/Dropbox/Sites/Learning/node/YDKJS/mario/src/components/static/imgs/speech-bubble.gif"
      />
      <img
        alt="Mario Charactor"
        className="Character__Mario-kSBfOl fcjCdJ"
        size={60}
        src="/Users/quochungle/Dropbox/Sites/Learning/node/YDKJS/mario/src/components/static/imgs/Mario.svg"
      />
    </div>
  </div>,
  <div
    className="Controls__Dashboard-cQakuF cCtKla"
  >
    <button
      className="Controls__Button-gzvyre dvIyxS"
      onClick={[Function]}
    >
      ← Move Left
    </button>
    <button
      className="Controls__Button-gzvyre dvIyxS"
      onClick={[Function]}
    >
      ↑ Jump
    </button>
    <button
      className="Controls__Button-gzvyre dvIyxS"
      onClick={[Function]}
    >
      Move Right →
    </button>
    <button
      className="Controls__Button-gzvyre dvIyxS"
      onClick={[Function]}
    >
      Toogle Labels
    </button>
  </div>,
]
`;
